Fix node 20 issue on GHA CI

Flamefire opened the pull request.

Flamefire committed:
Fix node 20 issue on GHA CI

joaquintides commented:
Hi @Flamefire, thanks for your contribution. Tests still fail for these jobs:

* https://github.com/boostorg/bimap/actions/runs/12215326386/job/34089119859?pr=47
* https://github.com/boostorg/bimap/actions/runs/12215326386/job/34089121257?pr=47
* https://github.com/boostorg/bimap/actions/runs/12215326386/job/34089121354?pr=47
* https://github.com/boostorg/bimap/actions/runs/12215326386/job/34089121505?pr=47

Any idea what's going on?

Flamefire committed:
Reformat CI and remove macos-12

Flamefire commented:
Some containers that don't need to be used. Reformatted to verify against the boost-ci template

Flamefire committed:
Fix clang toolset

Flamefire committed:
Install xz

Flamefire committed:
Remove C++2b from clang-17

There is a bug with libstdc++13 in 24.04: https://github.com/llvm/llvm-project/issues/59827

pdimov commented:
Why are you setting `toolset` to the wrong value (`toolset: clang++-3.9` instead of `toolset: clang-3.9`), and then patching it back in the script?

Flamefire committed:
Use clang-yy as toolset

Flamefire committed:
Bring back user config setting

Flamefire commented:
> Why are you setting `toolset` to the wrong value (`toolset: clang++-3.9` instead of `toolset: clang-3.9`), and then patching it back in the script?

Not really required in this repo indeed. That was from a RegEx replacement to ensure the correct compiler is used. IIRC I temporarily used `matrix.toolset` as the compiler in the user config but then decided against it.   
Changed
